File: frontend/src/types/payment.ts
// --- Enum types ---

export type PaymentOrderType =
  | "sepa_credit_transfer"
  | "riba"
  | "sdd"
  | "f24"
  | "bonifico_estero";

export type PaymentOrderStatus =
  | "bozza"
  | "da_approvare"
  | "approvata"
  | "file_generato"
  | "inviata_banca"
  | "eseguita"
  | "rifiutata"
  | "annullata";

export type PaymentPriority = "normale" | "alta" | "urgente";

export type PaymentLineStatus = "inclusa" | "eseguita" | "rifiutata" | "esclusa";

export type ApprovalAction = "approvata" | "rifiutata" | "richiesta_modifica";

export type SDDMandateType = "core" | "b2b";

export type SDDSequenceType = "frst" | "rcur" | "fnal" | "ooff";

export type F24SectionType = "erario" | "inps" | "regioni" | "imu_altri_enti";

// --- Label maps ---

export const PAYMENT_ORDER_TYPE_LABELS: Record<PaymentOrderType, string> = {
  sepa_credit_transfer: "Bonifico SEPA",
  riba: "RiBa",
  sdd: "SDD",
  f24: "F24",
  bonifico_estero: "Bonifico Estero",
};

export const PAYMENT_ORDER_STATUS_LABELS: Record<PaymentOrderStatus, string> = {
  bozza: "Bozza",
  da_approvare: "Da Approvare",
  approvata: "Approvata",
  file_generato: "File Generato",
  inviata_banca: "Inviata alla Banca",
  eseguita: "Eseguita",
  rifiutata: "Rifiutata",
  annullata: "Annullata",
};

export const PAYMENT_PRIORITY_LABELS: Record<PaymentPriority, string> = {
  normale: "Normale",
  alta: "Alta",
  urgente: "Urgente",
};

export const PAYMENT_LINE_STATUS_LABELS: Record<PaymentLineStatus, string> = {
  inclusa: "Inclusa",
  eseguita: "Eseguita",
  rifiutata: "Rifiutata",
  esclusa: "Esclusa",
};

export const F24_SECTION_LABELS: Record<F24SectionType, string> = {
  erario: "Erario",
  inps: "INPS",
  regioni: "Regioni",
  imu_altri_enti: "IMU / Altri enti",
};

// --- Status colors ---

export const PAYMENT_STATUS_COLORS: Record<PaymentOrderStatus, string> = {
  bozza: "bg-[#f2f2f2] text-[#817f7d]",
  da_approvare: "bg-[#ffd98c] text-[#1a1a1a]",
  approvata: "bg-[#c0c8ff] text-[#1a1a1a]",
  file_generato: "bg-[#d9deff] text-[#1a1a1a]",
  inviata_banca: "bg-[#c8b9df] text-[#1a1a1a]",
  eseguita: "bg-[#8cddba] text-[#1a1a1a]",
  rifiutata: "bg-[#fddcdc] text-[#ea0b0b]",
  annullata: "bg-[#e5e4e3] text-[#817f7d]",
};

// --- Interfaces ---

export interface PaymentOrderLine {
  id: string;
  payment_order_id: string;
  beneficiario_nome: string;
  beneficiario_iban: string | null;
  beneficiario_bic: string | null;
  beneficiario_cf: string | null;
  beneficiario_indirizzo: string | null;
  importo: number;
  valuta: string;
  causale: string | null;
  end_to_end_id: string | null;
  stato: PaymentLineStatus;
  counterparty_id: string | null;
  schedule_id: string | null;
  // RiBa
  riba_numero_ricevuta: string | null;
  riba_data_scadenza: string | null;
  riba_piazza: string | null;
  riba_abi_banca_domiciliataria: string | null;
  riba_cab_banca_domiciliataria: string | null;
  // SDD
  sdd_mandate_id: string | null;
  sdd_mandate_date: string | null;
  sdd_mandate_type: SDDMandateType | null;
  sdd_sequence_type: SDDSequenceType | null;
  // F24
  f24_sezione: F24SectionType | null;
  f24_codice_tributo: string | null;
  f24_rateazione: string | null;
  f24_anno_riferimento: number | null;
  f24_mese_riferimento: number | null;
  f24_importo_debito: number | null;
  f24_importo_credito: number | null;
  f24_codice_ente: string | null;
  f24_codice_sede: string | null;
  created_at: string;
  updated_at: string;
}

export interface PaymentApproval {
  id: string;
  payment_order_id: string;
  user_id: string;
  user_name: string | null;
  azione: ApprovalAction;
  commento: string | null;
  data_azione: string;
}

export interface PaymentOrder {
  id: string;
  company_id: string;
  bank_account_id: string | null;
  bank_account_nickname: string | null;
  tipo: PaymentOrderType;
  stato: PaymentOrderStatus;
  riferimento_interno: string;
  data_esecuzione_richiesta: string;
  importo_totale: number;
  numero_disposizioni: number;
  valuta: string;
  priorita: PaymentPriority;
  metadata_sepa: Record<string, unknown> | null;
  metadata_extra: Record<string, unknown> | null;
  file_generato_nome: string | null;
  file_generato_at: string | null;
  approvato_da_user_id: string | null;
  data_approvazione: string | null;
  creato_da_user_id: string | null;
  creato_da_user_name: string | null;
  note: string | null;
  lines: PaymentOrderLine[];
  created_at: string;
  updated_at: string;
}

export interface PaymentOrderSummary {
  totale_bozze: number;
  totale_da_approvare: number;
  totale_approvate: number;
  totale_eseguite: number;
  importo_bozze: number;
  importo_da_approvare: number;
  importo_approvate: number;
  importo_eseguite: number;
}

export interface PaymentOrderListResponse {
  items: PaymentOrder[];
  total: number;
  page: number;
  page_size: number;
  summary: PaymentOrderSummary;
}

export interface PaymentTemplate {
  id: string;
  company_id: string;
  nome: string;
  tipo: PaymentOrderType;
  bank_account_id: string | null;
  dati_template: Record<string, unknown> | null;
  ultimo_utilizzo: string | null;
  volte_utilizzato: number;
  created_at: string;
  updated_at: string;
}

export interface IbanValidationResponse {
  valid: boolean;
  iban_formatted: string | null;
  error: string | null;
}

// --- Workflow steps ---

export const WORKFLOW_STEPS: PaymentOrderStatus[] = [
  "bozza",
  "da_approvare",
  "approvata",
  "file_generato",
  "inviata_banca",
  "eseguita",
];

export const TERMINAL_STATUSES: PaymentOrderStatus[] = [
  "eseguita",
  "rifiutata",
  "annullata",
];

File: backend/app/schemas/payment.py
import uuid
from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, Field


# --- Response schemas ---

class PaymentOrderLineResponse(BaseModel):
    id: uuid.UUID
    payment_order_id: uuid.UUID
    beneficiario_nome: str
    beneficiario_iban: str | None = None
    beneficiario_bic: str | None = None
    beneficiario_cf: str | None = None
    beneficiario_indirizzo: str | None = None
    importo: Decimal
    valuta: str
    causale: str | None = None
    end_to_end_id: str | None = None
    stato: str
    counterparty_id: uuid.UUID | None = None
    schedule_id: uuid.UUID | None = None
    # RiBa
    riba_numero_ricevuta: str | None = None
    riba_data_scadenza: date | None = None
    riba_piazza: str | None = None
    riba_abi_banca_domiciliataria: str | None = None
    riba_cab_banca_domiciliataria: str | None = None
    # SDD
    sdd_mandate_id: str | None = None
    sdd_mandate_date: date | None = None
    sdd_mandate_type: str | None = None
    sdd_sequence_type: str | None = None
    # F24
    f24_sezione: str | None = None
    f24_codice_tributo: str | None = None
    f24_rateazione: str | None = None
    f24_anno_riferimento: int | None = None
    f24_mese_riferimento: int | None = None
    f24_importo_debito: Decimal | None = None
    f24_importo_credito: Decimal | None = None
    f24_codice_ente: str | None = None
    f24_codice_sede: str | None = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class PaymentApprovalResponse(BaseModel):
    id: uuid.UUID
    payment_order_id: uuid.UUID
    user_id: uuid.UUID
    user_name: str | None = None
    azione: str
    commento: str | None = None
    data_azione: datetime

    model_config = {"from_attributes": True}


class PaymentOrderResponse(BaseModel):
    id: uuid.UUID
    company_id: uuid.UUID
    bank_account_id: uuid.UUID | None = None
    bank_account_nickname: str | None = None
    tipo: str
    stato: str
    riferimento_interno: str
    data_esecuzione_richiesta: date
    importo_totale: Decimal
    numero_disposizioni: int
    valuta: str
    priorita: str
    metadata_sepa: dict | None = None
    metadata_extra: dict | None = None
    file_generato_nome: str | None = None
    file_generato_at: datetime | None = None
    approvato_da_user_id: uuid.UUID | None = None
    data_approvazione: datetime | None = None
    creato_da_user_id: uuid.UUID | None = None
    creato_da_user_name: str | None = None
    note: str | None = None
    lines: list[PaymentOrderLineResponse] = []
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class PaymentOrderSummaryResponse(BaseModel):
    totale_bozze: int = 0
    totale_da_approvare: int = 0
    totale_approvate: int = 0
    totale_eseguite: int = 0
    importo_bozze: Decimal = Decimal("0")
    importo_da_approvare: Decimal = Decimal("0")
    importo_approvate: Decimal = Decimal("0")
    importo_eseguite: Decimal = Decimal("0")


class PaymentOrderListResponse(BaseModel):
    items: list[PaymentOrderResponse]
    total: int
    page: int
    page_size: int
    summary: PaymentOrderSummaryResponse


class PaymentTemplateResponse(BaseModel):
    id: uuid.UUID
    company_id: uuid.UUID
    nome: str
    tipo: str
    bank_account_id: uuid.UUID | None = None
    dati_template: dict | None = None
    ultimo_utilizzo: datetime | None = None
    volte_utilizzato: int
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# --- Request schemas ---

class PaymentOrderCreateRequest(BaseModel):
    tipo: str
    bank_account_id: uuid.UUID | None = None
    data_esecuzione_richiesta: date
    valuta: str = "EUR"
    priorita: str = "normale"
    metadata_sepa: dict | None = None
    metadata_extra: dict | None = None
    note: str | None = None


class PaymentOrderUpdateRequest(BaseModel):
    bank_account_id: uuid.UUID | None = None
    data_esecuzione_richiesta: date | None = None
    valuta: str | None = None
    priorita: str | None = None
    metadata_sepa: dict | None = None
    metadata_extra: dict | None = None
    note: str | None = None


class PaymentOrderLineCreateRequest(BaseModel):
    beneficiario_nome: str = Field(min_length=1, max_length=255)
    beneficiario_iban: str | None = None
    beneficiario_bic: str | None = None
    beneficiario_cf: str | None = None
    beneficiario_indirizzo: str | None = None
    importo: Decimal = Field(gt=0)
    valuta: str = "EUR"
    causale: str | None = Field(None, max_length=140)
    end_to_end_id: str | None = Field(None, max_length=35)
    counterparty_id: uuid.UUID | None = None
    schedule_id: uuid.UUID | None = None
    # RiBa
    riba_numero_ricevuta: str | None = None
    riba_data_scadenza: date | None = None
    riba_piazza: str | None = None
    riba_abi_banca_domiciliataria: str | None = None
    riba_cab_banca_domiciliataria: str | None = None
    # SDD
    sdd_mandate_id: str | None = None
    sdd_mandate_date: date | None = None
    sdd_mandate_type: str | None = None
    sdd_sequence_type: str | None = None
    # F24
    f24_sezione: str | None = None
    f24_codice_tributo: str | None = None
    f24_rateazione: str | None = None
    f24_anno_riferimento: int | None = None
    f24_mese_riferimento: int | None = None
    f24_importo_debito: Decimal | None = None
    f24_importo_credito: Decimal | None = None
    f24_codice_ente: str | None = None
    f24_codice_sede: str | None = None


class PaymentOrderLineUpdateRequest(BaseModel):
    beneficiario_nome: str | None = Field(None, min_length=1, max_length=255)
    beneficiario_iban: str | None = None
    beneficiario_bic: str | None = None
    beneficiario_cf: str | None = None
    beneficiario_indirizzo: str | None = None
    importo: Decimal | None = Field(None, gt=0)
    valuta: str | None = None
    causale: str | None = Field(None, max_length=140)
    end_to_end_id: str | None = Field(None, max_length=35)
    counterparty_id: uuid.UUID | None = None
    schedule_id: uuid.UUID | None = None
    # RiBa
    riba_numero_ricevuta: str | None = None
    riba_data_scadenza: date | None = None
    riba_piazza: str | None = None
    riba_abi_banca_domiciliataria: str | None = None
    riba_cab_banca_domiciliataria: str | None = None
    # SDD
    sdd_mandate_id: str | None = None
    sdd_mandate_date: date | None = None
    sdd_mandate_type: str | None = None
    sdd_sequence_type: str | None = None
    # F24
    f24_sezione: str | None = None
    f24_codice_tributo: str | None = None
    f24_rateazione: str | None = None
    f24_anno_riferimento: int | None = None
    f24_mese_riferimento: int | None = None
    f24_importo_debito: Decimal | None = None
    f24_importo_credito: Decimal | None = None
    f24_codice_ente: str | None = None
    f24_codice_sede: str | None = None


class PaymentApprovalRequest(BaseModel):
    commento: str | None = None


class PaymentRejectRequest(BaseModel):
    commento: str = Field(min_length=1, max_length=1000)


class PaymentFromSchedulesRequest(BaseModel):
    schedule_ids: list[uuid.UUID] = Field(min_length=1)
    bank_account_id: uuid.UUID | None = None
    data_esecuzione_richiesta: date
    priorita: str = "normale"
    note: str | None = None


class IbanValidationRequest(BaseModel):
    iban: str = Field(min_length=1, max_length=34)


class IbanValidationResponse(BaseModel):
    valid: bool
    iban_formatted: str | None = None
    error: str | None = None


class PaymentTemplateCreateRequest(BaseModel):
    nome: str = Field(min_length=1, max_length=255)
    tipo: str
    bank_account_id: uuid.UUID | None = None
    dati_template: dict | None = None


class PaymentTemplateApplyRequest(BaseModel):
    data_esecuzione_richiesta: date
    override_bank_account_id: uuid.UUID | None = None
    override_priorita: str | None = None
    override_note: str | None = None


class MarkExecutedRequest(BaseModel):
    data_esecuzione: date | None = None
